mocks/log.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2015 ISRG.  All rights reserved
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package mocks
+
+import (
+	"log"
+	"log/syslog"
+	"regexp"
+
+	blog "github.com/letsencrypt/boulder/log"
+)
+
+// MockSyslogWriter implements the blog.SyslogWriter interface. It
+// stores all logged messages in a buffer for inspection by test
+// functions (via GetAll()) instead of sending them to syslog.
+type MockSyslogWriter struct {
+	logged    []*LogMessage
+	msgChan   chan<- *LogMessage
+	getChan   <-chan []*LogMessage
+	clearChan chan<- struct{}
+	closeChan chan<- struct{}
+}
+
+// LogMessage is a log entry that has been sent to a MockSyslogWriter.
+type LogMessage struct {
+	Priority syslog.Priority // aka Log level
+	Message  string          // content of log message
+}
+
+var levelName [8]string
+
+func init() {
+	levelName[syslog.LOG_EMERG] = "EMERG"
+	levelName[syslog.LOG_ALERT] = "ALERT"
+	levelName[syslog.LOG_CRIT] = "CRIT"
+	levelName[syslog.LOG_ERR] = "ERR"
+	levelName[syslog.LOG_WARNING] = "WARNING"
+	levelName[syslog.LOG_NOTICE] = "NOTICE"
+	levelName[syslog.LOG_INFO] = "INFO"
+	levelName[syslog.LOG_DEBUG] = "DEBUG"
+}
+
+func (lm *LogMessage) String() string {
+	return levelName[lm.Priority&7] + ": " + lm.Message
+}
+
+// SwitchLog changes the SyslogWriter used by the current singleton
+// audit logger, and returns the previous value. Example:
+//
+//	func TestFoo(t *testing.T) {
+//		log := mocks.NewSyslogWriter()
+//		defer log.Close()
+//		defer mocks.SwitchLog(mocks.SwitchLog(log))
+//		// ...
+//		Assert(t, len(log.GetAll()) > 0, "Should have logged something")
+//	}
+func SwitchLog(sw blog.SyslogWriter) blog.SyslogWriter {
+	al := blog.GetAuditLogger()
+	was := al.SyslogWriter
+	al.SyslogWriter = sw
+	return was
+}
+
+// UseMockLog changes the SyslogWriter used by the current singleton
+// audit logger to a new mock logger, and returns the mock. Example:
+//
+//	var log = mocks.UseMockLog()
+//	func TestFoo(t *testing.T) {
+//		log.Clear()
+//		// ...
+//		Assert(t, len(log.GetAll()) > 0, "Should have logged something")
+//	}
+func UseMockLog() *MockSyslogWriter {
+	sw := NewSyslogWriter()
+	blog.GetAuditLogger().SyslogWriter = sw
+	return sw
+}
+
+// NewSyslogWriter returns a new MockSyslogWriter.
+func NewSyslogWriter() *MockSyslogWriter {
+	msgChan := make(chan *LogMessage)
+	getChan := make(chan []*LogMessage)
+	clearChan := make(chan struct{})
+	closeChan := make(chan struct{})
+	msw := &MockSyslogWriter{
+		logged:    []*LogMessage{},
+		msgChan:   msgChan,
+		getChan:   getChan,
+		clearChan: clearChan,
+		closeChan: closeChan,
+	}
+	go func() {
+		for {
+			select {
+			case logMsg := <-msgChan:
+				log.Print("MockSyslog:" + logMsg.String())
+				msw.logged = append(msw.logged, logMsg)
```

Comment:
This is the only thread that reads the `logged` member. (I'm not just being pedantic: if you really mean "read `logged`", not "read the slice referenced by `logged`" then I think we're covered.) This thread does send the referenced slice over a channel, and the receiver can read the slice, but AFAIK it's safe for multiple goroutines to read from a slice. (Of course if multiple other goroutines start _modifying_ the slice elements, that's different.)

I do see a bug two lines later, though: it should be `msw.logged = []*LogMessage{}` so we don't overwrite the previous slice's backing array after sending it off to another goroutine.

Does this sound right, or am I still missing something?